core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -2345,17 +2345,19 @@ private[spark] class RedirectThread(
  */
 private[spark] class CircularBuffer(sizeInBytes: Int = 10240) extends java.io.OutputStream {
   var pos: Int = 0
+  var isBufferFull = false
   var buffer = new Array[Int](sizeInBytes)
 
   def write(i: Int): Unit = {
     buffer(pos) = i
     pos = (pos + 1) % buffer.length
+    isBufferFull = isBufferFull || (pos == 0)
   }
 
   override def toString: String = {
     val (end, start) = buffer.splitAt(pos)
     val input = new java.io.InputStream {
-      val iterator = (start ++ end).iterator
+      val iterator = (if (isBufferFull) start ++ end else end).iterator
```

Comment:
+1. I am happy to change that to make the codebase better. I am doing a string concat now, I thought of other ways but they all would end up with array copy inside String's constructor which the concat operation would also do.